### PR TITLE
 Variants Serial MPI openmp

### PR DESCRIPTION
**Pull request by akhilpatlolla**

### CODE REVIEW

File: var/spack/repos/builtin/packages/lulesh/package.py [hunk 1]
```diff
@@ -23,33 +23,59 @@
...
     url      = "https://codesign.llnl.gov/lulesh/lulesh2.0.3.tgz"
 
-    version("2.0.3", "336644a8750f71c7c6b9d2960976e7aa")
+    version('2.0.3', '336644a8750f71c7c6b9d2960976e7aa')
+    variant('mpi', default=True, description='Build with MPI support')
+    variant('openmp', default=True, description='Build with OpenMP support')
+    variant('visual', default=False,
+        description='Build with Visualization support (Silo, hdf5)')
 
-    patch("remove_defaults.patch")
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/lulesh/package.py]: If you're removing this patch, you should also remove the patch file.

File: var/spack/repos/builtin/packages/lulesh/package.py [hunk 2]
```diff
@@ -23,33 +23,59 @@
...
-    variant('mpip', default=False)
+    @property
+    def build_targets(self):
+        targets = []
+        cxxflag = ' -g -O3 -I. -Wall '
+        ldflags = ' -g -O3 '
 
-    depends_on("mpi", type="build")
-    depends_on("mpip", when="+mpip")
+        targets.append('CXX = {0}'.format(spack_cxx, ' -DUSE_MPI=0 '))
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/lulesh/package.py]: I don't think the string at the end is used in any way.

File: var/spack/repos/builtin/packages/lulesh/package.py [hunk 3]
```diff
@@ -23,33 +23,59 @@
...
+            cxxflag = ' -g -DVIZ_MESH -I${SILO_INCDIR} -Wall -Wno-pragmas '
+            ldflags = ' -g -L${SILO_LIBDIR} -Wl,-rpath -Wl,'
+            ldflags += '${SILO_LIBDIR} -lsiloh5 -lhdf5 '
 
-    def install(self, spec, prefix):
-        if '+mpip' in spec:
-            os.environ["LDFLAGS"] = " -lmpiP -ldwarf -lelf"
+        if '+mpi' in self.spec:
+            cxxflag += self.compiler.openmp_flag
+            ldflags += self.compiler.openmp_flag
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/lulesh/package.py]: Shouldn't these flags only be added when `+openmp`?

File: var/spack/repos/builtin/packages/lulesh/package.py [hunk 4]
```diff
@@ -23,33 +23,59 @@
...
-            if os.uname()[4] == "x86_64":
-                os.environ["LDFLAGS"] += " -lunwind"
+        targets.append('CXXFLAGS = {0}'.format(cxxflag))
+        targets.append('LDFLAGS = {0}'.format(ldflags))
+        return targets
 
-        os.environ["CXX"] = spec['mpi'].mpicxx + " -DUSE_MPI=1"
-        os.environ["PREFIX"] = prefix
-        make()
-        make("install")
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/lulesh/package.py]: The `install` target wasn't working?

File: var/spack/repos/builtin/packages/lulesh/package.py (hunk 2 above)
Comment (akhilpatlolla) [inline var/spack/repos/builtin/packages/lulesh/package.py]: It's specified in `makefile` ~

File: var/spack/repos/builtin/packages/lulesh/package.py [hunk 5]
```diff
@@ -23,33 +23,59 @@
...
+    def install(self, spec, prefix):
+        mkdirp(prefix.bin)
+        install('lulesh2.0', prefix.bin)
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/lulesh/package.py]: The name of this executable will probably change depending on the version. You should use `'lulesh{0}'.format(self.version.up_to(2))` here.

File: var/spack/repos/builtin/packages/lulesh/package.py [hunk 6]
```diff
@@ -23,33 +23,61 @@
...
     """
-
+    tags = ['proxy-app']
     homepage = "https://codesign.llnl.gov/lulesh.php"
     url      = "https://codesign.llnl.gov/lulesh/lulesh2.0.3.tgz"
 
-    version("2.0.3", "336644a8750f71c7c6b9d2960976e7aa")
-
-    patch("remove_defaults.patch")
+    version('2.0.3', '336644a8750f71c7c6b9d2960976e7aa')
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/lulesh/package.py]: It would be good to add an empty line here to visually separate the versions from the variants.

File: var/spack/repos/builtin/packages/lulesh/package.py [hunk 7]
```diff
@@ -23,33 +23,61 @@
...
 
-    variant('mpip', default=False)
+    depends_on('mpi', when='+mpi')
+    depends_on('silo', when='+visual')
+    depends_on('hdf5', when='+visual')
+    
+    @property
+    def build_targets(self):
+        targets = []
+        cxxflag = ' -g -O3 -I. -Wall '
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/lulesh/package.py]: Does `-Wall` work for every compiler?

File: var/spack/repos/builtin/packages/lulesh/package.py (hunk 2 above)
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/lulesh/package.py]: Yes, but what I'm saying is that the string formatting is going to use the first argument and ignore the second.

File: var/spack/repos/builtin/packages/lulesh/package.py [hunk 8]
```diff
@@ -23,33 +23,61 @@
...
+        ldflags = ' -g -O3 '
 
-    depends_on("mpi", type="build")
-    depends_on("mpip", when="+mpip")
+        targets.append('CXX = {0}'.format(spack_cxx, ' -DUSE_MPI=0 '))
+        targets.append('MPI_INC = {0}'.format(self.spec['mpi'].prefix.include))
+        targets.append('MPI_LIB = {0}'.format(self.spec['mpi'].prefix.lib))
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/lulesh/package.py]: You didn't check to see if `+mpi` is in the spec.

File: var/spack/repos/builtin/packages/lulesh/package.py [hunk 9]
```diff
@@ -23,33 +23,61 @@
...
+        if '+visual' in self.spec:
+            targets.append(
+                'SILO_INCDIR = {0}'.format(self.spec['silo'].prefix.include))
+            targets.append(
+                'SILO_LIBDIR = {0}'.format(self.spec['silo'].prefix.lib))
+            cxxflag = ' -g -DVIZ_MESH -I${SILO_INCDIR} -Wall -Wno-pragmas '
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/lulesh/package.py]: These flags definitely only work for GCC.

File: var/spack/repos/builtin/packages/lulesh/package.py [hunk 10]
```diff
@@ -23,33 +23,61 @@
...
+            ldflags = ' -g -L${SILO_LIBDIR} -Wl,-rpath -Wl,'
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/lulesh/package.py]: You don't need to bother with these, they are added by Spack's compiler wrappers.

File: var/spack/repos/builtin/packages/lulesh/package.py (hunk 9 above)
Comment (akhilpatlolla) [inline var/spack/repos/builtin/packages/lulesh/package.py]: Should I remove these flag then ?

Comment (adamjstewart) [inline var/spack/repos/builtin/packages/lulesh/package.py]: You could conditionally add them for GCC only if you want to. Depends on how involved you want to get. Ideally, any user could install this package with any compiler.

File: var/spack/repos/builtin/packages/lulesh/package.py [hunk 11]
```diff
@@ -23,33 +23,66 @@
...
+    @property
+    def build_targets(self):
+        targets = []
+        cxxflag = ' -g -O3 -I. '
+        ldflags = ' -g -O3 '
 
-    def install(self, spec, prefix):
-        if '+mpip' in spec:
-            os.environ["LDFLAGS"] = " -lmpiP -ldwarf -lelf"
+        targets.append('CXX = {0} {1}'.format(spack_cxx, ' -DUSE_MPI=0 '))
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/lulesh/package.py]: This should probably only be added when `'~mpi' in self.spec`. Otherwise you'll be setting `CXX` twice.

File: var/spack/repos/builtin/packages/lulesh/package.py [hunk 12]
```diff
@@ -23,33 +23,66 @@
...
+                'MPI_INC = {0}'.format(self.spec['mpi'].prefix.include))
+            targets.append('MPI_LIB = {0}'.format(self.spec['mpi'].prefix.lib))
+        if '+visual' in self.spec:
+            targets.append(
+                'SILO_INCDIR = {0}'.format(self.spec['silo'].prefix.include))
+            targets.append(
+                'SILO_LIBDIR = {0}'.format(self.spec['silo'].prefix.lib))
+            cxxflag = ' -g -DVIZ_MESH -I${SILO_INCDIR} '
+            if '+mpi' not in self.spec and '+openmp' not in self.spec:
+                cxxflag += ' -Wall -Wno-pragmas '
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/lulesh/package.py]: Why is this necessary? I don't think it will work with any non-GCC compilers.

File: var/spack/repos/builtin/packages/lulesh/package.py [hunk 13]
```diff
@@ -23,33 +23,66 @@
...
+            ldflags = ' -g -L${SILO_LIBDIR} -Wl,-rpath -Wl,'
+            ldflags += '${SILO_LIBDIR} -lsiloh5 -lhdf5 '
+
+        if '+openmp' in self.spec:
+            cxxflag += self.compiler.openmp_flag
+            ldflags += self.compiler.openmp_flag
 
-            if os.uname()[4] == "x86_64":
-                os.environ["LDFLAGS"] += " -lunwind"
+        if '+mpi' in self.spec:
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/lulesh/package.py]: Can you add this to the `if '+mpi' in self.spec` if-statement above? No need to check this variant multiple times.

File: var/spack/repos/builtin/packages/lulesh/package.py [hunk 14]
```diff
@@ -23,33 +23,62 @@
...
-    depends_on("mpi", type="build")
-    depends_on("mpip", when="+mpip")
+    @property
+    def build_targets(self):
+        targets = []
+        cxxflag = ' -g -O3 -I. '
+        ldflags = ' -g -O3 '
+        if '~mpi' in self.spec:
+            targets.append('CXX = {0} {1}'.format(spack_cxx, ' -DUSE_MPI=0 '))
+        if '+mpi' in self.spec:
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/lulesh/package.py]: It would be nicer to combine these:
```python
if '+mpi' in self.spec:
    ...
else:
    ...
```

File: var/spack/repos/builtin/packages/lulesh/package.py [hunk 15]
```diff
@@ -23,33 +23,62 @@
...
+            targets.append(
+                'MPI_INC = {0}'.format(self.spec['mpi'].prefix.include))
+            targets.append('MPI_LIB = {0}'.format(self.spec['mpi'].prefix.lib))
+        if '+visual' in self.spec:
+            targets.append(
+                'SILO_INCDIR = {0}'.format(self.spec['silo'].prefix.include))
+            targets.append(
+                'SILO_LIBDIR = {0}'.format(self.spec['silo'].prefix.lib))
+            cxxflag = ' -g -DVIZ_MESH -I${SILO_INCDIR} '
+            ldflags = ' -g -L${SILO_LIBDIR} -Wl,-rpath -Wl,'
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/lulesh/package.py]: This line is missing a space at the end. Possibly other lines as well. If you really want to be safe, it would be better to store these flags in a list and `' '.join(list)` them later.
